src/ring_theory/valuation/basic.lean

```diff
@@ -0,0 +1,392 @@
+{ to_fun := f ∘ v,
+  map_zero' := show f (v 0) = 0, by rw [v.map_zero, h₀],
+  map_add' := λ r s,
+    calc f (v (r + s)) ≤ f (max (v r) (v s))     : hf (v.map_add r s)
+                   ... = max (f (v r)) (f (v s)) : hf.map_max,
+  .. monoid_hom.comp f (v : R →* Γ₀) }
+
+/-- Two valuations on R are defined to be equivalent if they induce the same preorder on R. -/
+def is_equiv (v₁ : valuation R Γ₀) (v₂ : valuation R Γ'₀) : Prop :=
+∀ r s, v₁ r ≤ v₁ s ↔ v₂ r ≤ v₂ s
+
+end basic -- end of section
+
+namespace is_equiv
+variables [ring R]
+variables {v : valuation R Γ₀}
+variables {v₁ : valuation R Γ₀} {v₂ : valuation R Γ'₀} {v₃ : valuation R Γ''₀}
+
+@[refl] lemma refl : v.is_equiv v :=
+λ _ _, iff.refl _
+
+@[symm] lemma symm (h : v₁.is_equiv v₂) : v₂.is_equiv v₁ :=
+λ _ _, iff.symm (h _ _)
+
+@[trans] lemma trans (h₁₂ : v₁.is_equiv v₂) (h₂₃ : v₂.is_equiv v₃) : v₁.is_equiv v₃ :=
+λ _ _, iff.trans (h₁₂ _ _) (h₂₃ _ _)
+
+lemma of_eq {v' : valuation R Γ₀} (h : v = v') : v.is_equiv v' :=
+by subst h; refl
+
+-- move this
+lemma le_iff_of_mono_injective {α : Type*} {β : Type*} [linear_order α] [linear_order β] {f : α → β}
+  (f_mono : monotone f) (f_inj : injective f) {x y : α} :
+  x ≤ y ↔ f x ≤ f y :=
+⟨λ h, f_mono h,
+by { contrapose!, exact λ h, lt_of_le_of_ne (f_mono $ le_of_lt h) (f_inj.ne $ ne_of_lt h) }⟩
+
+lemma map {v' : valuation R Γ₀} (f : Γ₀ →* Γ'₀) (h₀ : f 0 = 0) (hf : monotone f) (inf : injective f)
+  (h : v.is_equiv v') :
+  (v.map f h₀ hf).is_equiv (v'.map f h₀ hf) :=
+λ r s,
+calc f (v r) ≤ f (v s) ↔ v r ≤ v s : by rw le_iff_of_mono_injective hf inf
+                   ... ↔ v' r ≤ v' s : h r s
+                   ... ↔ f (v' r) ≤ f (v' s) : by rw le_iff_of_mono_injective hf inf
+
+/-- `comap` preserves equivalence. -/
+lemma comap {S : Type u₂} [ring S] (f : S →+* R) (h : v₁.is_equiv v₂) :
+  (v₁.comap f).is_equiv (v₂.comap f) :=
+λ r s, h (f r) (f s)
+
+lemma val_eq (h : v₁.is_equiv v₂) {r s : R} :
+  v₁ r = v₁ s ↔ v₂ r = v₂ s :=
+⟨λ H, le_antisymm ((h _ _).1 (le_of_eq H)) ((h _ _).1 (le_of_eq H.symm)),
+ λ H, le_antisymm ((h.symm _ _).1 (le_of_eq H)) ((h.symm _ _).1 (le_of_eq H.symm)) ⟩
+
+lemma ne_zero (h : v₁.is_equiv v₂) {r : R} :
+  v₁ r ≠ 0 ↔ v₂ r ≠ 0 :=
+begin
+  have : v₁ r ≠ v₁ 0 ↔ v₂ r ≠ v₂ 0 := not_iff_not_of_iff h.val_eq,
+  rwa [v₁.map_zero, v₂.map_zero] at this,
+end
+
+end is_equiv -- end of namespace
+
+lemma is_equiv_of_map_strict_mono [ring R] {v : valuation R Γ₀}
+  (f : Γ₀ →* Γ'₀) (h₀ : f 0 = 0) (H : strict_mono f) :
+  is_equiv (v.map f h₀ (H.monotone)) v :=
+λ x y, ⟨H.le_iff_le.mp, λ h, H.monotone h⟩
+
+lemma is_equiv_of_val_le_one {K : Type*} [division_ring K]
+  (v : valuation K Γ₀) (v' : valuation K Γ'₀) (h : ∀ {x:K}, v x ≤ 1 ↔ v' x ≤ 1) :
+  v.is_equiv v' :=
+begin
+  intros x y,
+  by_cases hy : y = 0, { simp [hy, zero_iff], },
+  rw show y = 1 * y, by rw one_mul,
+  rw show x = (x * y⁻¹) * y, { rw [mul_assoc, inv_mul_cancel hy, mul_one], },
```

Comment:
Is this `inv_mul_cancel_assoc_left`? (side remark: that name is not very consistent with `inv_mul_cancel_right`)

Reply:
Fixed